mediagrains_async/__init__.py

```diff
@@ -0,0 +1,548 @@
+                meta['grain']['cog_coded_frame'] = {}
+                meta['grain']['cog_coded_frame']['format'] = await gbhd_child.read_uint(4)
+                meta['grain']['cog_coded_frame']['layout'] = await gbhd_child.read_uint(4)
+                meta['grain']['cog_coded_frame']['origin_width'] = await gbhd_child.read_uint(4)
+                meta['grain']['cog_coded_frame']['origin_height'] = await gbhd_child.read_uint(4)
+                meta['grain']['cog_coded_frame']['coded_width'] = await gbhd_child.read_uint(4)
+                meta['grain']['cog_coded_frame']['coded_height'] = await gbhd_child.read_uint(4)
+                meta['grain']['cog_coded_frame']['is_key_frame'] = await gbhd_child.read_bool()
+                meta['grain']['cog_coded_frame']['temporal_offset'] = await gbhd_child.read_sint(4)
+
+                if await gbhd_child.has_child_block():
+                    async with AsyncGSFBlock(self.file_data) as unof_block:
+                        meta['grain']['cog_coded_frame']['unit_offsets'] = []
+
+                        unit_offsets = await unof_block.read_uint(2)
+                        for i in range(0, unit_offsets):
+                            meta['grain']['cog_coded_frame']['unit_offsets'].append(await unof_block.read_uint(4))
+
+            elif gbhd_child.tag == "aghd":
+                meta['grain']['grain_type'] = "audio"
+                meta['grain']['cog_audio'] = {}
+                meta['grain']['cog_audio']['format'] = await gbhd_child.read_uint(4)
+                meta['grain']['cog_audio']['channels'] = await gbhd_child.read_uint(2)
+                meta['grain']['cog_audio']['samples'] = await gbhd_child.read_uint(4)
+                meta['grain']['cog_audio']['sample_rate'] = await gbhd_child.read_uint(4)
+
+            elif gbhd_child.tag == "cahd":
+                meta['grain']['grain_type'] = "coded_audio"
+                meta['grain']['cog_coded_audio'] = {}
+                meta['grain']['cog_coded_audio']['format'] = await gbhd_child.read_uint(4)
+                meta['grain']['cog_coded_audio']['channels'] = await gbhd_child.read_uint(2)
+                meta['grain']['cog_coded_audio']['samples'] = await gbhd_child.read_uint(4)
+                meta['grain']['cog_coded_audio']['priming'] = await gbhd_child.read_uint(4)
+                meta['grain']['cog_coded_audio']['remainder'] = await gbhd_child.read_uint(4)
+                meta['grain']['cog_coded_audio']['sample_rate'] = await gbhd_child.read_uint(4)
+
+            elif gbhd_child.tag == "eghd":
+                meta['grain']['grain_type'] = "event"
+            else:
+                raise GSFDecodeError(
+                    "Unknown type {} at offset {}".format(gbhd_child.tag, gbhd_child.block_start),
+                    gbhd_child.block_start,
+                    length=gbhd_child.size
+                )
+
+        return meta
+
+    async def decode_file_headers(self):
```

Comment:
Should this return `self.head` if it is not None? This would allow a client to get the header dict when using AsyncGSFDecoder as a context manager.

Reply:
Good call. WILCO